### DISCUSSION TITLE
Can I use 3D vars in certain fields?

### DISCUSSION THREAD

**Discussion by mahf708** [E3SM-Project/zppy#403] [Questions] [answered]
Hello team! Can I use a 3D variable in fields like `[[ atm_monthly_180x360_aave ]]`? I wanted to ask before submitting ... I suspect the answer is no, but I thought I would ask.

If a 3D variable is used in such a setting, how's that going to be plotted?

If not, what are the alternatives for 3D variables in the standard workflow?

**Comment by chengzhuzhang** [comment, marked as answer]
3D variable can be used in `[[ atm_monthly_180x360_aave ]]`
Plotting 3D variable is not included in current standard global time series plots. What type of plots are in mind? I think you could take the intermediate file generated by the [[ atm_monthly_180x360_aave ]] task and do the plotting offline at the time being.

**Comment by mahf708** [comment]
Yes, that makes perfect sense. I had nothing particular in mind, just thought about this while setting up some scripts and I thought I should know what is being done before attempting to run them and then waste time debugging. I will NOT include 3D variables for now, and stick to 2D variables only.

However, in the future, it could be beneficial to let the user select a specific level or average over a set of levels, etc. — though that may become cumbersome (and time consuming compute-wise). 

Thanks for the speedy response, marking your response as "Answer" 😃

**Reply by chengzhuzhang** [reply]
Hi @mahf708, I forgot to mention that for variables to feed into [[ atm_monthly_180x360_aave ]], It would be the best if you could include these default variables (see here, https://github.com/E3SM-Project/zppy/blob/c3f70dc8c814c36dcad5c380b1f86c8475a6192e/zppy/templates/default.ini#L49) These variables enables downstream E3SM Diags sets and overview global series plot. Among these variables, some are 3D, i.g. U is for the QBO diags in E3SM Ddiags.

**Reply by mahf708** [reply]
Thanks! I actually already ran into this very problem — it's like you saw my problems before me realizing them!! Naively, I thought I would get the same `vars` from the `ts` ones, but obviously not. See my script below and the corresponding link below; if you have any other ideas, please let me know :)


public [link](https://compy-dtn.pnl.gov/mahf708/E3SMv3integration/20230129.AWG_fixedSST_sensitivity.NGD_v3atm.2xDUST.compy/e3sm_diags/atm_monthly_180x360_aave_mvm/model_vs_model_2004-2013/viewer/)

<details> <summary> zppy script </summary>

```cfg
[default]
input = /compyfs/mahf708/E3SMv3integration/20230129.AWG_fixedSST_sensitivity.NGD_v3atm.2xDUST.compy
output = /compyfs/mahf708/E3SMv3integration/20230129.AWG_fixedSST_sensitivity.NGD_v3atm.2xDUST.compy
case = 20230129.AWG_fixedSST_sensitivity.NGD_v3atm.2xDUST.compy
www = /compyfs/www/mahf708/E3SMv3integration
partition = short
environment_commands = "source /share/apps/E3SM/conda_envs/load_latest_e3sm_unified_compy.sh"

[climo] 
active = True
walltime = "1:00:00"
years ="2004:2013:10",

  [[ atm_monthly_180x360_aave ]]
  input_subdir = "archive/atm/hist"
  input_files = "eam.h0"
  mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  frequency = "monthly"

  [[ atm_monthly_diurnal_8xdaily_180x360_aave ]]
  input_subdir = "archive/atm/hist"
  input_files = "eam.h4"
  mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  vars = "PRECT"
  frequency = "diurnal_8xdaily"

[ts]
active = True
walltime = "0:30:00"
years ="2004:2013:5",

  [[ atm_monthly_180x360_aave ]]
  input_subdir = "archive/atm/hist"
  input_files = "eam.h0"
  mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  frequency = "monthly"
  vars = "FSNTOA,FLUT,FSNT,FSNT_d1,FSNTC,FSNTC_d1,FLNT,FLNT_d1,FLNTC,FLNTC_d1,FSNS,FLNS,SHFLX,QFLX,TAUX,TAUY,PRECC,PRECL,PRECSC,PRECSL,TS,TREFHT,CLDTOT,CLDHGH,CLDMED,CLDLOW,U,ICEFRAC,LANDFRAC,OCNFRAC,AODALL,AODBC,AODDUST,AODPOM,AODSO4,AODSOA,AODSS,AODVIS,PS,BURDENSEASALT,BURDENSO4,BURDENDUST"

  # [[ atm_monthly_180x360_aave ]]
  # input_subdir = "archive/atm/hist"
  # input_files = "eam.h7"
  # mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  # frequency = "monthly"
  # vars = "cdnc,lwp,iwp,lcc,icc,clt,cod,ccn,ttop,OMEGA500,OMEGA700,TH7001000,U850,V850,SOLIN,FSNT,FSNTOA,FSUTOA,FSUTOA_d1,FSUTOAC,FSUTOAC_d1,FLUT,FLNT,FLUTC,FLNTC,FSNSC,FSDSC,CLDLOW_CAL,CLDMED_CAL,CLDHGH_CAL,AODVIS,FSNT,FLNT,FSNTC,FLNTC,FSNT_d1,FLNT_d1,FSNTC_d1,FLNTC_d1,FSNS,FLNS,FSNSC,FLNSC,FSNS_d1,FLNS_d1,FSNSC_d1,FLNSC_d1,CLDHGH,CLDMED,CLDLOW,T,TS,TREFHT,BURDEN1,BURDEN2,BURDEN3,BURDEN4,BURDEN5,BURDENSO4,BURDENSEASALT,BURDENDUST,AODSS,AODSO4,AODDUST"

  [[ atm_daily_180x360_aave ]]
  input_subdir = "archive/atm/hist"
  input_files = "eam.h1"
  frequency = "daily"
  mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  vars = "PRECT"

  [[ atm_monthly_glb ]]
  input_subdir = "archive/atm/hist"
  input_files = "eam.h0"
  frequency = "monthly"
  mapping_file = "glb"

  [[ land_monthly ]]
  input_subdir = "archive/lnd/hist"
  input_files = "elm.h0"
  frequency = "monthly"
  mapping_file = /qfs/people/zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc
  vars = "FSH,RH2M"
  extra_vars = "landfrac"

  [[ rof_monthly ]]
  input_subdir = "archive/rof/hist"
  input_files = "mosart.h0"
  mapping_file = ""
  frequency = "monthly"
  vars = "RIVER_DISCHARGE_OVER_LAND_LIQ"
  extra_vars = 'areatotal2'

[e3sm_diags]
active = True
years = "2004:2013:10"
ts_num_years = 5
ref_start_yr = 1979
ref_final_yr = 2016

  [[ atm_monthly_180x360_aave ]]
  short_name = '20230129.AWG_fixedSST_sensitivity.NGD_v3atm.2xDUST.compy'
  grid = '180x360_aave'
  sets = 'lat_lon','zonal_mean_xy','zonal_mean_2d','polar','cosp_histogram','meridional_mean_2d','enso_diags','qbo','diurnal_cycle','annual_cycle_zonal_mean','streamflow','zonal_mean_2d_stratosphere'
  reference_data_path = '/compyfs/e3sm_diags_data/obs_for_e3sm_diags/climatology'
  obs_ts = '/compyfs/e3sm_diags_data/obs_for_e3sm_diags/time-series'
  dc_obs_climo = '/compyfs/e3sm_diags_data/obs_for_e3sm_diags/climatology'
  climo_diurnal_subsection = "atm_monthly_diurnal_8xdaily_180x360_aave"
  climo_diurnal_frequency = "diurnal_8xdaily"
  output_format_subplot = "pdf",

  [[ atm_monthly_180x360_aave_mvm ]]
  climo_subsection = "atm_monthly_180x360_aave"
  diff_title = "Difference"
  grid = '180x360_aave'
  ref_final_yr = 2004
  ref_start_yr = 2013
  reference_data_path = '/compyfs/mahf708/E3SMv3integration/20230129.AWG_fixedSST_sensitivity.NGD_v3atm.baseline.compy/post/atm/180x360_aave/clim/'
  ref_name = "20230129.AWG_fixedSST_sensitivity.NGD_v3atm.baseline.compy"
  ref_years = "2004-2013",
  run_type = "model_vs_model"
  years = "2004-2013", 
  short_ref_name = 'baseline'
  short_name = '2xDUST'
  swap_test_ref = False
  tag = "model_vs_model"
  ts_num_years_ref = 5
  ts_subsection = "atm_monthly_180x360_aave"

[global_time_series]
active = True
atmosphere_only = True
years = "2004-2013",
ts_num_years = 5
figstr= "2xDUST"
experiment_name = "2xDUST"
ts_years ="2004-2013", 
climo_years = "2004-2013", 
```

</details>

**Reply by mahf708** [reply]
Btw, I noticed two things that I should ask you about:
- Sometimes I see `plev` or `lev` in these settings, does that mean one can select a given level for these plots or is it better handled at the namelist level? I don't really have specific ones of interest now because the obvious ones are available as 2D plots (think omega500 and so on) but just in case it is good to know of already builtin functionality for future analysis
- I sometimes see `vars` and `extra_vars` — what's going on; how do they differ? Also, I assume I can have duplicate of these blocks with different variables, right?

**Comment by mahf708** [comment]
Btw @chengzhuzhang since I have you here, I have another question that's been bugging me! Why do I get my raw output files saved under `${CASE_ROOT}/run/` whereas I see others have their files under `${CASE_ROOT}/archive/atm/hist/` etc. I would like to have my files follow the latter structure. Any ideas what I should do to get this done?

**Comment by chengzhuzhang** [comment]
Thanks for noticing this and raising the question. I think that `short term archiving` was not turned on in your simulations. But it is strongly recommended, see here: https://acme-climate.atlassian.net/wiki/spaces/ED/pages/2309226536/Running+E3SM+step-by-step+guide#Short-Term-Archiving.

**Reply by mahf708** [reply]
Thank you so much! This is all I needed precisely because of the reasons listed in the page.

**Reply by chengzhuzhang** [reply]
I think it is worthwhile to point this guide (thanks to @forsyth2 ) to everyone new to E3SM :)

**Reply by forsyth2** [reply]
Yes, thanks @chengzhuzhang! I tried to cover as much as possible in that guide :)

**Reply by mahf708** [reply]
Can I comment on that document? I found that using `./case.st_archive --last-date 0051-01-01 --force-move --no-incomplete-logs` was scary-ish, so I instead used `./case.st_archive`. Just a thought for completed simulations. The user may think they have to use the extra flags...

**Reply by forsyth2** [reply]
Yes, you should be able to add comments on any Confluence page.

**Reply by mahf708** [reply]
Btw, the automated st archive (`DO_SHORT_TERM_ARCHIVING=true) doesn't work, at least not for me. The manual one works fine. Do you know where I should report this issue or whom to ask about it?

**Reply by forsyth2** [reply]
Not sure why this is the case. Also posted in response to your question on Slack.
